From 1e41f86e4f979ece07b1e97c1036d4043853021f Mon Sep 17 00:00:00 2001
From: Zong Li <zong.li@sifive.com>
Date: Thu, 11 Mar 2021 17:28:32 +0800
Subject: [PATCH 51/52] board: sifive: Add an interface to get PCB revision

There are different DDR parameter settings for different board revisions,
so add an new interface to get the PCB revision to determine which DDR
parameter should be used.

Signed-off-by: Zong Li <zong.li@sifive.com>
---
 arch/riscv/include/asm/arch-fu740/eeprom.h    | 15 +++++++++
 .../hifive-platform-i2c-eeprom.c              | 32 +++++++++++++++++++
 2 files changed, 47 insertions(+)
 create mode 100644 arch/riscv/include/asm/arch-fu740/eeprom.h

diff --git a/arch/riscv/include/asm/arch-fu740/eeprom.h b/arch/riscv/include/asm/arch-fu740/eeprom.h
new file mode 100644
index 0000000000..0e1220e558
--- /dev/null
+++ b/arch/riscv/include/asm/arch-fu740/eeprom.h
@@ -0,0 +1,15 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (C) 2021 SiFive, Inc.
+ *
+ * Zong Li <zong.li@sifve.com>
+ */
+
+#ifndef _ASM_RISCV_EEPROM_H
+#define _ASM_RISCV_EEPROM_H
+
+#define PCB_REVISION_REV3	0x3
+
+u8 get_pcb_revision_from_eeprom(void);
+
+#endif /* _ASM_RISCV_EEPROM_H */
diff --git a/board/sifive/hifive_unmatched_fu740/hifive-platform-i2c-eeprom.c b/board/sifive/hifive_unmatched_fu740/hifive-platform-i2c-eeprom.c
index 851a31378e..5770390ddf 100644
--- a/board/sifive/hifive_unmatched_fu740/hifive-platform-i2c-eeprom.c
+++ b/board/sifive/hifive_unmatched_fu740/hifive-platform-i2c-eeprom.c
@@ -534,3 +534,35 @@ int mac_read_from_eeprom(void)
 
 	return 0;
 }
+
+/**
+ * get_pcb_revision_from_eeprom - get the PCB revision
+ *
+ * Read the EEPROM to determine the board revision.
+ *
+ * This function is called before relocation, so we need to read a private
+ * copy of the EEPROM into a local variable on the stack.
+ */
+u8 get_pcb_revision_from_eeprom(void)
+{
+	struct __attribute__ ((__packed__)) board_eeprom {
+		u8 magic[MAGIC_NUMBER_BYTES];
+		u8 format_ver;
+		u16 product_id;
+		u8 pcb_revision;
+	} be;
+
+	int ret;
+	struct udevice *dev;
+
+	ret = i2c_get_chip_for_busnum(CONFIG_SYS_EEPROM_BUS_NUM,
+				      CONFIG_SYS_I2C_EEPROM_ADDR,
+				      1,
+				      &dev);
+
+	if (!ret)
+		ret = dm_i2c_read(dev, 0, (void *)&be,
+				  sizeof(struct board_eeprom));
+
+	return be.pcb_revision;
+}
-- 
2.30.2

